src/common/translate_item.js

```diff
@@ -105,8 +105,52 @@ Zotero.Translate.ItemSaver.prototype = {
 			payload.cookie = document.cookie;
 		}
 		payload.proxy = this._proxy && this._proxy.toJSON();
+		
+		// Add page data for snapshot
+		// TODO: Check for skipSnapshot
+		let singleFile;
+		for (let i = 0; i < items.length; i++) {
+			for (let j = 0; j < items[i].attachments.length; j++) {
+				if (items[i].attachments[j].mimeType !== 'text/html'
+					|| items[i].attachments[j].snapshot === false) {
+					continue;
+				}
+
+				// If we have not done so already, inject SingleFile and retrieve
+				// the page content
+				if (!singleFile) {
+					singleFile = await Zotero.SingleFile.retrievePageData();
```

Comment:
This can't be the same logic as saveSnapshot, because in this case there's no guarantee that it's the current page. We only want to retrieve page data when the attachment URL matches the current URL.

Also, this can now only throw, not return undefined, so, as above, this should be in a try/catch and fall back to saving in the client if possible.

Reply:
Ok. I think I had the URL comparison originally, but it was breaking on twitter because the query param was removed. Is that the only modification the attachment URL will have? I guess I could match on either the full or without the query params. Are those URLs coming from the translator?

Reply:
Hmm, this is tricky. OK, so to review the problem: Previously, the translator assigned a URL for each attachment, and the client would just retrieve that URL (with passed cookies) and save. We now want to save the current page in the connector if possible, but there's not really a great way to know if a current query string is meaningful and it should match the attachment URL.

We probably want to say that if 1) there's a single item being saved and 2) the full attachment URL matches some variation of the current URL, we assume the current page is the snapshot. Easy enough to match on full current URL or current URL with query string and fragment removed, but technically if the attachment URL has a query string we should match regardless of query parameter order, which would mean normalizing both the attachment URL and current URL to alphabetize any query parameters and remove any fragment.

What was the query param you were seeing on Twitter?

We should still fall back to the client for saving non-matching URLs (including for most multiples), so if our logic is imperfect here and we save the current page via the client with cookies, that's OK (though it won't work saving to server).

Reply:
I think everything you said makes perfect sense! The query param I saw was `lang=en`, on twitter and probably found via Google. My biggest worry here is saving files when we are not supposed to.

I went back and double checked what was going on. I was saving a twitter page using Embedded Metadata and currently the ItemSaver comparing the `item.url` to `attachment.url`. I was incorrect though. `item.url` is coming across without the query parameter, but the `attachment.url` has the query param, at least in this example.

But it's not too complicated to build in the logic you suggested and I can't see a way that will go wrong so I'll go ahead and do that. And I'll compare against `location.href` not `item.url`.